src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -79,62 +80,68 @@ public T runWithAnyNode() {
+      } catch (JedisConnectionException e) {
+        connectionSupplier = handleConnectionProblem(slot, currentAttempt);
+      } catch (JedisRedirectionException e) {
+        redirectionSupplier = handleRedirection(connection, e);
+      } finally {
+        releaseConnection(connection);
       }
+    }
 
-      return execute(connection);
-
-    } catch (JedisNoReachableClusterNodeException jnrcne) {
-      throw jnrcne;
-    } catch (JedisConnectionException jce) {
-      // release current connection before recursion
-      releaseConnection(connection);
-      connection = null;
-
-      if (attempts <= 1) {
-        //We need this because if node is not reachable anymore - we need to finally initiate slots
-        //renewing, or we can stuck with cluster state without one node in opposite case.
-        //But now if maxAttempts = [1 or 2] we will do it too often.
-        //TODO make tracking of successful/unsuccessful operations for node - do renewing only
-        //if there were no successful responses from this node last few seconds
-        this.connectionHandler.renewSlotCache();
-      }
+    throw new JedisClusterMaxAttemptsException("No more cluster attempts left.");
+  }
 
-      return runWithRetries(slot, attempts - 1, tryRandomNode, redirect);
-    } catch (JedisRedirectionException jre) {
-      // if MOVED redirection occurred,
-      if (jre instanceof JedisMovedDataException) {
-        // it rebuilds cluster's slot cache recommended by Redis cluster specification
-        this.connectionHandler.renewSlotCache(connection);
-      }
+  private Supplier<Jedis> handleConnectionProblem(final int slot, int currentAttempt) {
+    int attemptsLeft = (maxAttempts - currentAttempt) - 1;
+    if (attemptsLeft <= 1) {
+      //We need this because if node is not reachable anymore - we need to finally initiate slots
+      //renewing, or we can stuck with cluster state without one node in opposite case.
+      //But now if maxAttempts = [1 or 2] we will do it too often.
+      //TODO make tracking of successful/unsuccessful operations for node - do renewing only
+      //if there were no successful responses from this node last few seconds
+      this.connectionHandler.renewSlotCache();
+    }
 
-      // release current connection before recursion
-      releaseConnection(connection);
-      connection = null;
+    return () -> connectionHandler.getConnectionFromSlot(slot);
+  }
 
-      return runWithRetries(slot, attempts - 1, false, jre);
-    } finally {
-      releaseConnection(connection);
+  private Supplier<Jedis> handleRedirection(Jedis connection, final JedisRedirectionException jre) {
+    // if MOVED redirection occurred,
+    if (jre instanceof JedisMovedDataException) {
+      // it rebuilds cluster's slot cache recommended by Redis cluster specification
+      this.connectionHandler.renewSlotCache(connection);
     }
+
+    // release current connection before iteration
+    releaseConnection(connection);
```

Comment:
```suggestion
```

Reply:
In the previous code, release connection can solve a connection leak, see:https://github.com/redis/jedis/pull/1005 
but i don't think we need it, because we do not have recursion, and `releaseConnection ` in `finally`.